mono/sgen/sgen-marksweep.c

```diff
@@ -1870,17 +1870,18 @@ sgen_evacuation_freelist_blocks (MSBlockInfo * volatile *block_list, int size_in
 }
 
 static void
-major_start_major_collection (void)
+major_start_major_collection (gboolean compacting)
 {
 	MSBlockInfo *block;
 	int i;
 
 	major_finish_sweep_checking ();
 
-	/*
-	 * Clear the free lists for block sizes where we do evacuation.  For those block
-	 * sizes we will have to allocate new blocks.
-	 */
+	if (compacting) {
```

Comment:
This is a bad heuristic, it will double the heap size while compacting.

It should take into account how much headroom we have left when figuring out how much to compact.